crates/viewer/re_space_view_spatial/src/contexts/transform_context.rs

```diff
@@ -254,51 +347,130 @@ impl TransformContext {
         &self.space_origin
     }
 
-    /// Retrieves the transform of on entity from its local system to the space of the reference.
+    /// Retrieves transform information for a given entity.
     ///
-    /// Returns None if the path is not reachable.
-    pub fn reference_from_entity(&self, ent_path: &EntityPath) -> Option<glam::Affine3A> {
-        self.transform_per_entity
-            .get(ent_path)
-            .map(|i| i.reference_from_entity)
+    /// Returns `None` if it's not reachable from the view's origin.
+    pub fn transform_info_for_entity(&self, ent_path: &EntityPath) -> Option<&TransformInfo> {
+        self.transform_per_entity.get(ent_path)
     }
+}
 
-    /// Like [`Self::reference_from_entity`], but if `ent_path` has a pinhole camera, it won't affect the transform.
-    ///
-    /// Normally, the transform we compute for an entity with a pinhole transform places all objects
-    /// in front (defined by view coordinates) of the camera with a given image plane distance.
-    /// In some cases like drawing the lines for a frustum or arrows for the 3D transform, this is not the desired transformation.
-    /// Returns None if the path is not reachable.
-    ///
-    /// TODO(#2663, #1025): Going forward we should have separate transform hierarchies for 2D (i.e. projected) and 3D,
-    /// which would remove the need for this.
-    pub fn reference_from_entity_ignoring_pinhole(
-        &self,
-        ent_path: &EntityPath,
-        entity_db: &EntityDb,
-        query: &LatestAtQuery,
-    ) -> Option<glam::Affine3A> {
-        let transform_info = self.transform_per_entity.get(ent_path)?;
-        if let (true, Some(parent)) = (
-            transform_info.parent_pinhole.as_ref() == Some(ent_path),
-            ent_path.parent(),
+/// Compute transform info for when we walk up the tree from the reference.
+fn transform_info_for_upward_propagation(
+    reference_from_ancestor: glam::Affine3A,
+    transforms_at_entity: TransformsAtEntity,
+) -> TransformInfo {
+    let mut reference_from_entity = reference_from_ancestor;
+
+    // Need to take care of the fact that we're walking the other direction of the tree here compared to `transform_info_for_downward_propagation`!
+    // Apply inverse transforms in flipped order!
```

Comment:
Very much share this concern. I think if it wasn't for 2D/3D handling we should do this differently: Instead of having per space view transform contexts, we could instead have a singel transform tree for each 3D space - typically just a single space starting at '/'.
Then, instead of working with different camera references, we can manipulate the eye camera matrix (or if we want to be more expressive, have another transform stage we could call "origin", making our conceptual pipeline `instance -> entity -> world -> origin -> camera -> projection`) such that it appears that $origin is the actual origin of the space. This way we don't have to do any inverse transforms at all.

I'll open an issue about this :thinking:

Reply:
https://github.com/rerun-io/rerun/issues/7025